File: backend/app/schemas/epaper.py
from __future__ import annotations

from datetime import date, datetime, time
from typing import Literal

from pydantic import BaseModel, Field


class EpaperArticleOut(BaseModel):
    id: int
    short_id: str
    url: str
    title_te: str
    title_en: str | None = None
    summary_te: str | None = None
    hero_url: str | None = None
    category_slug: str | None = None
    is_breaking: bool = False
    audio_url: str | None = None
    position: int
    display_type: str


class EpaperPageOut(BaseModel):
    id: int
    page_number: int
    title: str
    layout_type: str
    share_url: str
    articles: list[EpaperArticleOut]
    poll_id: int | None = None


class EpaperEditionOut(BaseModel):
    id: int
    title: str
    edition_date: date
    edition_type: str
    status: str
    revision: int
    page_count: int
    pages: list[EpaperPageOut] = []
    pdf_url: str | None = None
    audio_enabled: bool = False
    published_at: datetime | None = None


class PageTemplateIn(BaseModel):
    slug: str = Field(pattern=r"^[a-z0-9][a-z0-9-]*$", max_length=100)
    title_te: str = Field(min_length=2, max_length=180)
    title_en: str = Field(min_length=2, max_length=180)
    sort: int = Field(default=0, ge=0, le=200)
    category_ids: list[int] = Field(default_factory=list, max_length=30)
    story_count: int = Field(default=6, ge=1, le=20)
    layout_type: Literal[
        "lead_grid", "two_column", "three_column", "image_lead", "briefs", "breaking"
    ] = "lead_grid"
    is_visible: bool = True


class GenerateEditionIn(BaseModel):
    edition_date: date | None = None


class PageUpdateIn(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=180)
    layout_type: str | None = Field(default=None, max_length=30)
    article_ids: list[int] | None = Field(default=None, max_length=20)
    poll_id: int | None = None


class PageCreateIn(BaseModel):
    title: str = Field(min_length=2, max_length=180)
    layout_type: Literal[
        "lead_grid", "two_column", "three_column", "image_lead", "briefs", "breaking"
    ] = "lead_grid"
    article_ids: list[int] = Field(default_factory=list, max_length=20)
    poll_id: int | None = None


class PageOrderIn(BaseModel):
    page_ids: list[int] = Field(min_length=1, max_length=50)


PreferenceType = Literal["category", "district", "mandal", "tag"]


class UserEditionPreferenceIn(BaseModel):
    preference_type: PreferenceType
    target_id: int = Field(gt=0)
    priority: int = Field(default=0, ge=0, le=100)


class UserEditionIn(BaseModel):
    name: str = Field(min_length=2, max_length=160)
    auto_generate: bool = True
    generation_time: time = time(6, 0)
    preferences: list[UserEditionPreferenceIn] = Field(min_length=1, max_length=30)


class UserEditionOut(BaseModel):
    id: int
    name: str
    auto_generate: bool
    generation_time: time
    is_active: bool
    preferences: list[UserEditionPreferenceIn]
    latest_edition: EpaperEditionOut | None = None

File: frontend/src/types/epaper.ts
export interface EpaperArticle {
  id:number; short_id:string; url:string; title_te:string; title_en:string|null;
  summary_te:string|null; hero_url:string|null; category_slug:string|null;
  is_breaking:boolean; audio_url:string|null; position:number; display_type:string;
}
export interface EpaperPage { id:number; page_number:number; title:string; layout_type:string; share_url:string; articles:EpaperArticle[]; poll_id:number|null }
export interface EpaperEdition { id:number; title:string; edition_date:string; edition_type:string; status:string; revision:number; page_count:number; pages:EpaperPage[]; pdf_url:string|null; audio_enabled:boolean; published_at:string|null }
export interface EpaperArchive { items:EpaperEdition[] }
export interface AudioTrack { article_id:number; short_id:string; title_te:string; url:string; page_number:number }
export interface EpaperAudio { enabled:boolean; tracks:AudioTrack[] }
export interface UserEditionPreference { preference_type:'category'|'district'|'mandal'|'tag'; target_id:number; priority:number }
export interface UserEdition { id:number; name:string; auto_generate:boolean; generation_time:string; is_active:boolean; preferences:UserEditionPreference[] }
export interface PollOption { id:number; option_text_te:string; option_text_en:string|null; votes:number; percentage:number }
export interface Poll { id:number; question_te:string; question_en:string|null; status:string; is_big_question:boolean; start_time:string; end_time:string; category_id:number|null; district_id:number|null; article_id:number|null; total_votes:number; has_voted:boolean; selected_option_id:number|null; options:PollOption[] }
export interface TrendingTopic { slug:string; title_te:string; title_en:string|null; score:number; is_override:boolean }
